File: app/src/main/java/me/xiaopan/android/spear/sample/fragment/ImageFragment.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@xiaopan.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.android.spear.sample.fragment;

import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import me.xiaoapn.android.spear.sample.R;
import me.xiaopan.android.spear.request.DisplayListener;
import me.xiaopan.android.spear.request.ProgressCallback;
import me.xiaopan.android.spear.sample.DisplayOptionsType;
import me.xiaopan.android.spear.sample.widget.ProgressPieView;
import me.xiaopan.android.spear.util.FailureCause;
import me.xiaopan.android.spear.widget.SpearImageView;

public class ImageFragment extends Fragment {
	public static final String PARAM_REQUIRED_IMAGE_URI = "PARAM_REQUIRED_IMAGE_URI";
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		String uri = null;
		Bundle bundle = getArguments();
		if(bundle != null){
			uri = bundle.getString(PARAM_REQUIRED_IMAGE_URI);
		}
		if(uri != null){
			View rootView = inflater.inflate(R.layout.fragment_image, container, false);
			final ProgressPieView progressBar = (ProgressPieView) rootView.findViewById(R.id.progress_imageFragment_progress);
            progressBar.setMax(100);

            SpearImageView imageView = (SpearImageView) rootView.findViewById(R.id.image_imageFragment_image);
			imageView.setDisplayOptions(DisplayOptionsType.VIEW_PAGER);
            imageView.setDisplayListener(new DisplayListener() {
                @Override
                public void onStarted() {
                    progressBar.setVisibility(View.VISIBLE);
                    progressBar.setProgress(0);
                }

                @Override
                public void onFailed(FailureCause failureCause) {
                    progressBar.setVisibility(View.GONE);
                }

                @Override
                public void onCompleted(String imageUri, ImageView imageView, BitmapDrawable drawable, From from) {
                    progressBar.setVisibility(View.GONE);
                }

                @Override
                public void onCanceled() {
                }
            });
            imageView.setProgressCallback(new ProgressCallback() {
                @Override
                public void onUpdateProgress(long totalLength, long completedLength) {
                    progressBar.setProgress((int) (((float) completedLength / totalLength) * 100));
                }
            });
            imageView.setImageByUri(uri);
			return rootView;
		}else{
			return null;
		}
	}
}

File: sample/src/androidMain/kotlin/com/github/panpf/sketch/sample/ui/test/ImageFragment.kt
/*
 * Copyright (C) 2022 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.sample.ui.test

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.navigation.fragment.navArgs
import com.github.panpf.assemblyadapter.pager.FragmentItemFactory
import com.github.panpf.sketch.cache.CachePolicy.DISABLED
import com.github.panpf.sketch.displayImage
import com.github.panpf.sketch.request.LoadState.Error
import com.github.panpf.sketch.sample.databinding.FragmentImageBinding
import com.github.panpf.sketch.sample.ui.model.ImageDetail
import com.github.panpf.sketch.sample.ui.base.BaseBindingFragment
import com.github.panpf.sketch.sample.ui.util.createThemeSectorProgressDrawable
import com.github.panpf.sketch.sample.util.repeatCollectWithLifecycle
import com.github.panpf.sketch.util.SketchUtils
import com.github.panpf.sketch.ability.showDataFromLogo
import com.github.panpf.sketch.ability.showProgressIndicator

class ImageFragment : BaseBindingFragment<FragmentImageBinding>() {

    private val args by navArgs<ImageFragmentArgs>()

    override var isPage = false

    override fun onViewCreated(binding: FragmentImageBinding, savedInstanceState: Bundle?) {
        binding.myImage.apply {
            showDataFromLogo()
            showProgressIndicator(createThemeSectorProgressDrawable(requireContext()))
            displayImage(args.url) {
                memoryCachePolicy(DISABLED)
                resultCachePolicy(DISABLED)
                downloadCachePolicy(DISABLED)
            }
        }

        binding.smallState.apply {
            binding.myImage.requestState.loadState
                .repeatCollectWithLifecycle(viewLifecycleOwner, Lifecycle.State.STARTED) {
                    if (it is Error) {
                        error {
                            retryAction {
                                SketchUtils.restart(binding.myImage)
                            }
                        }
                    } else {
                        gone()
                    }
                }
        }
    }

    class ItemFactory : FragmentItemFactory<ImageDetail>(ImageDetail::class) {

        override fun createFragment(
            bindingAdapterPosition: Int,
            absoluteAdapterPosition: Int,
            data: ImageDetail
        ): Fragment = ImageFragment().apply {
            arguments = ImageFragmentArgs(
                data.mediumUrl,
            ).toBundle()
        }
    }
}